draft-ietf-quic-load-balancers.md

```diff
@@ -449,66 +449,52 @@ The nonce length MUST be no fewer than 4 octets.
 The configuration agent also selects an 16-octet AES-ECB key to use for
 connection ID decryption.
 
-### Load Balancer Actions {#encrypted-short-load-balancer-actions}
+### Server Actions
 
-Upon receipt of a QUIC packet, the load balancer extracts as many of the
-earliest octets from the destination connection ID as necessary to match the
-server ID. The nonce immediately follows.
+When generating a routable connection ID, the server writes arbitrary bits into
+its nonce octets, and its provided server ID into the server ID octets. See
+{{cid-entropy}} for nonce generation considerations.
 
-The load balancer decrypts the nonce and the server ID using the following three
-pass algorithm:
+The server encrypts the server ID using the following four pass algorithm, which
+leverages 128-bit AES Electronic Codebook (ECB) mode, much like QUIC header
+protection.
 
-* Pass 1:  The load balancer decrypts the server ID using 128-bit AES Electronic
-Codebook (ECB) mode, much like QUIC header protection. The encrypted nonce
-octets are zero-padded to 16 octets.  AES-ECB encrypts this encrypted nonce
-using its key to generate a mask which it applies to the encrypted server id.
-This provides an intermediate value of the server ID, referred to as server-id
-intermediate.
+In the text below, ^ is the XOR function, || is concatenation, and "zeros" is of
+length equal to 15 octets minus the server ID length, so that the input to
+AES-ECB is 16 octets.
 
-server_id_intermediate = encrypted_server_id ^ AES-ECB(key, padded-encrypted-nonce)
+The truncate() function takes the most significant octets of its argument, so that
+the XOR function operates on fields of the same length.
 
-* Pass 2:  The load balancer decrypts the nonce octets using 128-bit AES
-ECB mode, using the server-id intermediate as "nonce" for this pass. The
-server-id intermediate octets are zero-padded to 16 octets.  AES-ECB encrypts
-this padded server-id intermediate using its key to generate a mask which it
-applies to the encrypted nonce. This provides the decrypted nonce value.
+First, set server_id_0 to the server ID, and nonce_0 to the nonce.
 
-nonce = encrypted_nonce ^ AES-ECB(key, padded-server_id_intermediate)
+Then apply the following steps:
 
-* Pass 3:  The load balancer decrypts the server ID using 128-bit AES ECB mode.
-The nonce octets are zero-padded to 16 octets.  AES-ECB encrypts this nonce
-using its key to generate a mask which it applies to the intermediate server id.
-This provides the decrypted server ID.
+    server_id_1 = server_id_0 ^ (truncate(AES-ECB(key, nonce_0 || zeros || 0x01)))
+    nonce_1 = nonce_0 ^ truncate(AES-ECB(key, server_id_1 || zeros || 0x02))
+    server_id_2 = server_id_1 ^ (truncate(AES-ECB(key, nonce_1 || zeros || 0x03)))
+    nonce_2 = nonce_1 ^ (truncate(AES-ECB(key, server_id_2 || zeros || 0x04)))
 
-server_id = server_id_intermediate ^ AES-ECB(key, padded-nonce)
+The encrypted CID is ```first_octet || server_id_2 || nonce_2```.
```

Comment:
I would process in a different order:

    nonce1 = nonce_0 ^ truncate(AES-ECB(key, server_id_0 || zeros || 0x01))
    server_id_1= server_id_0 ^ (truncate(AES-ECB(key, nonce_1 || zeros || 0x02)))
    nonce_2 = nonce_1 ^ (truncate(AES-ECB(key, server_id_1 || zeros || 0x03)))
    server_id_2 = server_id_1 ^ (truncate(AES-ECB(key, nonce_2 || zeros || 0x02)))

This has a couple of advantages:

1) There is a risk that servers will use a counter instead of a random number for the nonce. Scrambling the nonce in the first pass makes sense there.
2) The first pass can be pre-computed just once per server-id
3) If you are not interested in the decrypted value of the nonce, then the server ID is decrypted in just three passes.

But then, there is also a point about using left and right halves instead of splitting at server_id and nonce.

Reply:
1. The point of the design is that the server SHOULD use a counter; that's how we shrank the nonce! I'm not sure why the order here matters as long as we encrypt both sides
2 and 3 are good points; I'll update this PR, but I'm curious what the crypto review folks would say.

Reply:
I see your point now about balancing the size of each pass. I'll do that now.

Reply:
An alternative would be to ask each server implementation to pick a random "nonce mask" for each encryption period, and to construct the nonce as "counter xor mask". That would have the same effect as the first pass that I proposed, and you could then keep your original proposal. But you would need four passes for decryption.

Reply:
The order matters because the counter values used as nonce are predictable. I am not exactly a crypto expert, but I am pretty sure that predictable values are easier to crack than random values.

Reply:
You may be right, but a major part of this design is to allow a counting nonce, so that it can be smaller without triggering the birthday problem.

Perhaps if we just encouraged the service to start with a random first nonce value?

Reply:
OK, I suggested a random initial nonce

Reply:
Yes, that's a good suggestion, can only help.